odoo/cli/obfuscate.py

```diff
@@ -0,0 +1,205 @@
+
+    def run(self, cmdargs):
+        parser = odoo.tools.config.parser
+        group = optparse.OptionGroup(parser, "Populate Configuration")
+        group.add_option('--pwd', dest="pwd", default=False, help="Cypher password")
+        group.add_option('--fields', dest="fields", default=False, help="List of table.columns to obfuscate/unobfuscate: table1.column1,table2.column1,table2.column2")
+        group.add_option('--exclude', dest="exclude", default=False, help="List of table.columns to exclude from obfuscate/unobfuscate: table1.column1,table2.column1,table2.column2") 
+        group.add_option('--unobfuscate', action='store_true', default=False)
+        group.add_option('--vacuum', action='store_true', default=False, help="Vacuum database after unobfuscating")
+        group.add_option('--pertablecommit', action='store_true', default=False, help="Commit after each table instead of a big transaction")
+
+        parser.add_option_group(group)
+        if not cmdargs:
+            sys.exit(parser.print_help())
+
+        try:
+            opt = odoo.tools.config.parse_config(cmdargs)
+            dbname = odoo.tools.config['db_name']
+            self.registry = odoo.registry(dbname)
+            with self.registry.cursor() as cr:
+                self.cr = cr
+                self.begin()
+                if self.check_pwd(opt.pwd):
+                    fields = [
+                        ('mail_tracking_value', 'old_value_char'),
+                        ('mail_tracking_value', 'old_value_text'),
+                        ('mail_tracking_value', 'new_value_char'),
+                        ('mail_tracking_value', 'new_value_text'),
+                        ('res_partner', 'name'),
+                        ('res_partner', 'display_name'),
+                        ('res_partner', 'email'),
+                        ('res_partner', 'phone'),
+                        ('res_partner', 'mobile'),
+                        ('res_partner', 'street'),
+                        ('res_partner', 'street2'),
+                        ('res_partner', 'city'),
+                        ('res_partner', 'zip'),
+                        ('res_partner', 'vat'),
+                        ('res_partner', 'website'),
+                        ('res_country', 'name'),
+                        ('mail_message', 'subject'),
+                        ('mail_message', 'email_from'),
+                        ('mail_message', 'reply_to'),
+                        ('mail_message', 'body'),
+                        ('crm_lead', 'title'),
+                        ('crm_lead', 'name'),
+                        ('crm_lead', 'contact_name'),
+                        ('crm_lead', 'partner_name'),
+                        ('crm_lead', 'email_from'),
+                        ('crm_lead', 'phone'),
+                        ('crm_lead', 'mobile'),
+                        ('crm_lead', 'website'),
+                        ('crm_lead', 'description'),
+                    ]
```

Comment:
This looks not very extendable.
I think we need to think about all the app store modules that add obfuscable fields.

**Proposal 1** : create a dedicated overloadable function.

**proposal 2** : Add an option on the field definition. Something like : 

```
class ResPartner(Model):
    _name = "res.partner"

    phone = fields.Char(obfuscate=True)
```

As a developper, I think that the proposal 2 is better. when I develop a new module, I will ask myself if the field contains confidential data when writing the field ( = fields()...)

Reply:
Proposal 2 they won't accept in stable versions I guess :-) But yes we need more flexibility/freedom.

Reply:
> Proposal 2 they won't accept in stable versions I guess

I'm not sure to see why could be unstable in adding an extra new args obfuscate (defaut False) in fields.
It doesn't change the db layout. Maybe I missed something. Could you elaborate ?

Reply:
at least, the `--fields` option is here to add fields

Reply:
> at least, the --fields option is here to add fields

Hi @nseinlet. Do you see any inconvenience in switching to a ``fields.Char(obfusctate=True)`` syntax ? What is your point of view regarging this proposal ? 

My worry about the current implementation is that if any developper add a new field that has to be obfuscated (odoo core team, or odoo partners, in custom modules), people that run the cli tools will not know that they have to add a new item in --fields argument.

Reply:
I think datas can be retrieve from elsewhere in the application, like task descriptions, attachments, ... It's the admin responsibility to know what (s)he does, and a way to achieve this is to explicitely ask them a list of fields.

Let's say you mark all the sensitive fields in your module as `obfuscate=True`, and a custom field `x_...` is a related stored field to yours, what is the chance for an admin to miss it?

Reply:
Hi @nseinlet. 
I definitively don't understand your point. I don't ask to remove the option to obfuscate some extra fields. I think that instead of writting an hardcoded list of field in that file, it's better to define that a field is obfuscated in the field definition itself.
Of course, if a user with high privilege create a new custom field ("x_...") with sensitive data, the admin user should identify it. (but the same occures with the current design).

Regarding your example of custom field (x_...) related to stored field, it can be set as obfuscated automatically. (I mean a related field of an obfuscated field should be obfuscated). Same control occures with fields type. (a field.integer can not be related to a field.Text for exemple).

If you don't like the syntax ``name = fields.Char(obfuscate=True)`` could you, at least move the hardcoded list in a dedicated overloadable function. 
Something in BaseModel, like ``def _get_obfucated_fields()`` that returns the current list, and that can be overloaded to add extra fields.

@etobella : If I remember correctly, you implemented Odoo in some hospitals, with I guess a lot of data to be obfuscated. Maybe you'll have a point of view on that topic ? 

Thanks for your attention.

Reply:
If we can avoid modifying the ORM in a stable version, it is better.

Reply:
@legalsylvain @etobella I don't think this feature is secure enough for sending hospital sensitive data.

Reply:
Don't worry, we were using OpenUpgrade for security concerns

Reply:
> Don't worry, we were using OpenUpgrade for security concerns

I know and it's great !
 
My question was more "what about a database with a lot of extra fields (non-core) to be obfuscated" what do you think about the current implementation of this PR ?

Reply:
This tool doesn't use the ORM or registry. You can specify fields in a file.

Reply:
I don't get it. line 139 there is ``            self.registry = odoo.registry(dbname)``

Anyway, if the new option "obfuscate" is present as a new column in the ir_model_fields table, the list can be generated by SQL.

What do you think ?

thanks !

Reply:
We can't add a column in stable version.
And I'm not fan of this idea of new field property.

Reply:
>  And I'm not fan of this idea of new field property.

Well, that's not a valid argument. The current design you propose doesn't work for customized database. (=database with modules that are not Odoo CE / EE).
Consider a complex situation : An odoo instance, with some extra modules installed from the appstore, some extra custom modules developped by many developpers for many years, and some extra fields added by UI, if you ask to a people to send you the database obfuscated, he can NOT know which extra fields he has to obfuscate and he has no easy way to define the list correctly. (In my instance, I have 965 tables, and > 10000 columns)
Obfuscated fields should be flagged at developpement / implementation step. not later.

To be sure, I think you can ask the administrator of the odoo instance (odoo.com) to provide you an obfuscated database. (I understand you have some custom code on your instance.)
I'm sure it'll take him a long time to define the list of extra fields to obfuscate, and there'll be some leakage of personal data.

In addition, having a new attribute at field level makes it possible to meet needs that are not currently covered, such as having an option to hide personal data during live demonstrations, etc...

Reply:
> he can NOT know

This is the point. if the db admin doesn't know, the admin probably miss something because a field is not marked as "to obfuscate". Having the admin to provide a list , the admin is aware of what he's doing.

> To be sure, I think you can ask the administrator of the odoo instance (odoo.com) to provide you an obfuscated database.

I'm aware of the fields to obfuscate. And I' m also aware this obfuscation is not secure.

The aim of this is to provide a database for upgrade. But some fields needs to be processed during upgrade, and thus cannot be obfuscated. Which ones? it depends on the source and target versions. So it's not a silver bullet.

Reply:
@nseinlet I'm not sure to understand how adding an attribute to a field would hinder stability. It would be an optional flag, no harm if not present. And you can surely init an env in your tool to check which filelds are enabled for obfuscation.

You assume a db admin knows everything about the db... but IMO that's wrong. That might be true for one-man-show projects. When you work on more than 20/30/40 projects at time, in a team of 20/30/40 ppl you cannot have the full knowledge and you cannot rely on the fact that whoever is doing the tech work pre-migration knows what has to be obfuscated or not. 

IMO it would be really nice if you can define programmatically what has to be obfuscated and what not w/o having to maintain a list of table/fields and/or additional files are your DB evolves.

Maybe I'm missing how do this tool is intended to be used though...
Could you please add a proper description to the PR explaining what led to the current implementation?
Merci! :pray: 

My $0.02 :)